astacus/node/cassandra.py

```diff
@@ -0,0 +1,156 @@
+                ipc.CassandraSubOp.remove_snapshot: self.remove_snapshot,
+                ipc.CassandraSubOp.restore_snapshot: self.restore_snapshot,
+                ipc.CassandraSubOp.start_cassandra: self.start_cassandra,
+                ipc.CassandraSubOp.stop_cassandra: self.stop_cassandra,
+                ipc.CassandraSubOp.take_snapshot: self.take_snapshot,
+            }[subop],
+        )
+
+    def _get_schema_hash(self):
+        with CassandraClient(self.config.cassandra.client).connect() as cas:
+            schema = CassandraSchema.from_cassandra_session(cas)
+            return schema.calculate_hash()
+
+    def get_schema_hash(self):
+        """This is used to get hash of the schema as seen by this node."""
+        with contextlib.suppress(TransientException):
+            self.result.schema_hash = self._get_schema_hash()
+        self.result.progress.done()
+
+    def remove_snapshot(self):
+        """This is used to remove the current snapshot (if any).
+
+        It is used as prelude for actual Astacus snapshot of the files
+        and after the backup has completed.
+
+        Note that Cassandra does not do any internal bookkeeping of
+        the snapshots so the rmtrees are enough.
+        """
+        progress = self.result.progress
+        progress.add_total(1)
+        todo = list(self.config.root.glob(SNAPSHOT_GLOB))
+        progress.add_success()
+        progress.add_total(len(todo))
+        for snapshotpath in todo:
+            shutil.rmtree(snapshotpath)
+            progress.add_success()
+        progress.done()
+
+    def restore_snapshot(self):
+        """This is used to restore the snapshot files into place, with Cassandra offline."""
+        # TBD: Delete extra data (current cashew doesn't do it, but we could)
+
+        # Move files from Astacus snapshot directories to the actual data directories
+        progress = self.result.progress
+        table_snapshots = list(self.config.root.glob(SNAPSHOT_GLOB))
+        progress.add_total(len(table_snapshots))
+
+        for table_snapshot in table_snapshots:
+            parts = table_snapshot.parts
+            # -2 = snapshots, -1 = name of the snapshots
+            table_name_and_id = parts[-3]
+            keyspace_name = parts[-4]
+            if is_system_keyspace(keyspace_name):
+                progress.add_success()
+                continue
+
+            table_name, _ = table_name_and_id.rsplit("-", 1)
+
+            # This could be more efficient too; oh well.
+            keyspace_path = table_snapshot.parents[2]
+            table_paths = list(keyspace_path.glob(f"{table_name}-*"))
+            assert len(table_paths) >= 1, f"NO tables with prefix {table_name}- found in {keyspace_path}!"
+            if len(table_paths) > 1:
+                # Prefer the one that isn't table_name_and_id
+                table_paths = [p for p in table_paths if p.name != table_name_and_id]
```

Comment:
wait - there were cases when `p.name == table_name_and_id`? I kinda thought that if you execute `CREATE TABLE` again on the fresh cluster you get a different ID...

Reply:
This is looking at the snapshot (and table_name_and_id is from snapshot too). In some cases it seems same table name can have more than one subdirectory; not sure if this logic is copied from Cashew or new though.

Reply:
I don't know what is the history behind this either anymore :-) This is actually probably artifact of restoring stuff locally to same cluster.

What is your suggestions on what to do with it - get rid of it?

Reply:
just checked, Cassandra does not clean up old table directory when you do DROP TABLE. Odd. So the multiple dir case is more common than I thought, I though you could get that only as a race condition when creating from different nodes or something.

So we need to be more clever, unfortunately. The table ID is in `system.schema`:
```
aiven@cqlsh> SELECT id FROM system_schema.tables WHERE keyspace_name='foo' AND table_name='bar';

 id
--------------------------------------
 77567990-e801-11ec-8abe-e178ff91d9b8
```
but not in client metadata, it seems. Oh well. We can include table ID in the manifest and restore data from the version active at the time we took the backup.

Reply:
By default there are none though - the restore to non-empty case is simply something for later. Note: If it is not in metadata, it is quite awkward to rely on it as I think the content of system tables is not guaranteed to stay same across versions.